firmware/library/L2_HAL/sensors/temperature/temperature.hpp

```diff
@@ -10,3 +11,69 @@ class TemperatureInterface
   virtual float GetFahrenheit() = 0;
 };
 
+class Temperature : public TemperatureInterface
+{
+ public:
+  static constexpr uint8_t kTemperatureAddress      = 0x31;
+  static constexpr uint8_t kTemperatureIdRegister   = 0xC0;
+  static constexpr uint8_t kMemoryMapAddress        = 0x14;
+  static constexpr uint8_t kOneBurstRegister        = 0xC4;
+  static constexpr uint8_t kAutonicBitRegister      = 0xC5;
+  static constexpr uint8_t kMostSignificantRegister = 0xC1;
+
+  constexpr Temperature() : i2c_(&lpc40xx_i2c_) {}
+
+  explicit Temperature(I2cInterface * lpc40xx_i2c_select)
+      : i2c_(lpc40xx_i2c_select)
+  {
+  }
+
+  bool Initialize() override
+  {
+    i2c_->Initialize();
+    uint8_t temperature_sensor_id_register;
+    // is the same as the Base ID (0x14).
+    i2c_->WriteThenRead(kTemperatureAddress, &kTemperatureIdRegister, 1,
+                        &temperature_sensor_id_register, 1);
+    return (temperature_sensor_id_register == kMemoryMapAddress);
+  }
+
+  float GetCelsius() override
+  {
+    uint8_t most_significant_register;
+    uint8_t least_significant_register;
+    // The register will enable the device to collect data once
+    // and automatically sets the stop bit to 0 (2nd bit).
+    i2c_->Write(kTemperatureAddress, { kOneBurstRegister, 0x04 }, 2);
+    // Auto increments I2c register address pointer.
+    i2c_->Write(kTemperatureAddress, { kAutonicBitRegister, 0x01 }, 2);
+    i2c_->WriteThenRead(kTemperatureAddress, &kMostSignificantRegister, 1,
+                        &most_significant_register, 1);
+    i2c_->Read(kTemperatureAddress, &least_significant_register, 1);
+
+    uint16_t temperature_data = 0;
```

Comment:
As I said offline, you need to make sure this variable is set to 0 since the clear and Insert function preserve the previous bits, meaning anything that wasn't touched is still undefined.